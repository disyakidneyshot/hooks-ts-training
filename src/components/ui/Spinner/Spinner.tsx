import React from 'react'
import {
	Wrapper,
	BallsContainer,
	FirstBall,
	SecondBall,
	ThirdBall,
} from './styledComps'

export interface ISpinnerProps {
	size?: number
}

export const Spinner: React.FC<ISpinnerProps> = (props): JSX.Element => {
	return (
		<Wrapper {...props}>
			<BallsContainer {...props}>
				<FirstBall {...props} />
				<SecondBall {...props} />
				<ThirdBall {...props} />
			</BallsContainer>
		</Wrapper>
	)
}

Spinner.defaultProps = {
	size: 25,
}
