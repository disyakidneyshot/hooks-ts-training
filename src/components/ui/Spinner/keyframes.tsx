import { keyframes } from 'styled-components'

export const spinnerMovement = (size: number) => keyframes`
0% {
  opacity: 1;
  transform: translate(0 0);
}
49.99% {
  opacity: 1;
  transform: translate(${size}px, 0);
}
50% {
  opacity: 0;
  transform: translate(${size}px, 0);
}
100% {
  opacity: 0;
  transform: translate(0, 0);
}
`

export const spinnerMovementRev = (size: number) => keyframes`
0% {
    transform: translate(0, 0);
  }
50% {
  transform: translate(${size}px, 0);
}
100% {
  transform: translate(0, 0);
}
`
