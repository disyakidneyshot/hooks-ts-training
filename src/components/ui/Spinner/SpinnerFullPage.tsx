import React from 'react'
import { Spinner } from './Spinner'
import styled from 'styled-components'

const SpinnerContainer = styled.div`
	width: 100vw;
	height: 100vh;
	align-items: center;
	justify-content: center;
`

export const FullPageSpinner: React.FC = (): JSX.Element => {
	return (
		<SpinnerContainer>
			<Spinner />
		</SpinnerContainer>
	)
}
