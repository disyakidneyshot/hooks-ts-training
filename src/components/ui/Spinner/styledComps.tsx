import styled from 'styled-components'
import { spinnerMovement, spinnerMovementRev } from './keyframes'
import { ISpinnerProps } from './Spinner'

export const Wrapper = styled.div<ISpinnerProps>`
	width: ${props => props.size && props.size * 2}px;
	height: ${props => props.size && props.size}px;
	display: inline-block;
	overflow: hidden;
`

export const BallsContainer = styled.div<ISpinnerProps>`
	width: 100%;
	height: 100%;
	position: relative;
	transform: translateZ(0) scale(1);
	backface-visibility: hidden;
	transform-origin: 0 0;
`

export const Ball = styled.div<ISpinnerProps>`
	position: absolute;
	width: ${props => props.size && props.size}px;
	height: ${props => props.size && props.size}px;
	border-radius: 50%;
	top: 0px;
	left: 0px;
	box-sizing: content-box;
`

export const FirstBall = styled(Ball)<ISpinnerProps>`
	background: ${props => props.theme.success};
	animation: ${props => props.size && spinnerMovementRev(props.size)} 1s linear
		infinite;
	animation-delay: -0.5s;
`
export const SecondBall = styled(Ball)<ISpinnerProps>`
	background: ${props => props.theme.accent};
	animation: ${props => props.size && spinnerMovementRev(props.size)} 1s linear
		infinite;
	animation-delay: 0s;
`

export const ThirdBall = styled(Ball)<ISpinnerProps>`
	background: ${props => props.theme.success};
	animation: ${props => props.size && spinnerMovement(props.size)} 1s linear
		infinite;
	animation-delay: -0.5s;
`
