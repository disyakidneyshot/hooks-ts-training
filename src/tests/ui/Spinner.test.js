import React from 'react'
import { shallow, mount } from 'enzyme'
import 'jest-styled-components'
import { Spinner } from '../../components/ui/Spinner/Spinner'

describe('Component: <Spinner />', () => {
	const sizeProp = 23
	const Wrapper = shallow(<Spinner size={sizeProp} />)

	it('renders without crashing', () => {
		expect(Wrapper).toMatchSnapshot()
	})

	it('components size is based on prop:size', () => {
		const MounterWrapper = mount(<Spinner size={sizeProp} />)
		const Ball = MounterWrapper.find('Ball').first()
		expect(MounterWrapper).toHaveStyleRule('width', `${sizeProp * 2}px`)
		expect(MounterWrapper).toHaveStyleRule('height', `${sizeProp}px`)
		expect(Ball).toHaveStyleRule('width', `${sizeProp * 2}px`)
		expect(Ball).toHaveStyleRule('height', `${sizeProp}px`)
	})
})
