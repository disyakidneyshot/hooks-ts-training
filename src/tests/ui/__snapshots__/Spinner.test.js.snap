// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Spinner /> renders without crashing 1`] = `
<styled.div
  size={23}
>
  <styled.div
    size={23}
  >
    <Styled(styled.div)
      size={23}
    />
    <Styled(styled.div)
      size={23}
    />
    <Styled(styled.div)
      size={23}
    />
  </styled.div>
</styled.div>
`;
